File: services/clinical-service/src/domain/aggregates/Encounter.ts
import { v4 as uuidv4 } from 'uuid';
import { ClinicalNote, ClinicalNoteProps } from '@domain/entities/ClinicalNote';
import { Diagnosis } from '@domain/value-objects/Diagnosis';
import { DomainEvent } from '@domain/events/EncounterCreatedEvent';
import { EncounterCreatedEvent } from '@domain/events/EncounterCreatedEvent';
import { PatientAdmittedEvent } from '@domain/events/PatientAdmittedEvent';
import { PatientDischargedEvent } from '@domain/events/PatientDischargedEvent';

export enum EncounterStatus {
    PLANNED = 'PLANNED',
    IN_PROGRESS = 'IN_PROGRESS',
    COMPLETED = 'COMPLETED',
    CANCELLED = 'CANCELLED',
}

export interface EncounterProps {
    id?: string;
    patientId: string;
    clinicalNotes?: ClinicalNote[];
    diagnoses?: Diagnosis[];
    status?: EncounterStatus;
    startDate?: Date;
    endDate?: Date | null;
    createdAt?: Date;
    updatedAt?: Date;
}

export class Encounter {
    readonly id: string;
    readonly patientId: string;
    private _clinicalNotes: ClinicalNote[];
    private _diagnoses: Diagnosis[];
    private _status: EncounterStatus;
    readonly startDate: Date;
    private _endDate: Date | null;
    readonly createdAt: Date;
    private _updatedAt: Date;

    private _domainEvents: DomainEvent[] = [];

    private constructor(props: Required<Omit<EncounterProps, 'endDate'>> & { endDate: Date | null }) {
        this.id = props.id;
        this.patientId = props.patientId;
        this._clinicalNotes = props.clinicalNotes;
        this._diagnoses = props.diagnoses;
        this._status = props.status;
        this.startDate = props.startDate;
        this._endDate = props.endDate;
        this.createdAt = props.createdAt;
        this._updatedAt = props.updatedAt;
    }

    static create(props: EncounterProps): Encounter {
        const encounter = new Encounter({
            id: props.id ?? uuidv4(),
            patientId: props.patientId,
            clinicalNotes: props.clinicalNotes ?? [],
            diagnoses: props.diagnoses ?? [],
            status: props.status ?? EncounterStatus.PLANNED,
            startDate: props.startDate ?? new Date(),
            endDate: props.endDate ?? null,
            createdAt: props.createdAt ?? new Date(),
            updatedAt: props.updatedAt ?? new Date(),
        });

        encounter.addDomainEvent(
            new EncounterCreatedEvent(encounter.id, encounter.patientId)
        );

        return encounter;
    }

    // ── Getters ──

    get status(): EncounterStatus {
        return this._status;
    }

    get endDate(): Date | null {
        return this._endDate;
    }

    get clinicalNotes(): ReadonlyArray<ClinicalNote> {
        return [...this._clinicalNotes];
    }

    get diagnoses(): ReadonlyArray<Diagnosis> {
        return [...this._diagnoses];
    }

    get updatedAt(): Date {
        return this._updatedAt;
    }

    // ── Domain behaviour ──

    admit(): void {
        if (this._status !== EncounterStatus.PLANNED) {
            throw new Error('Only planned encounters can be admitted');
        }

        this._status = EncounterStatus.IN_PROGRESS;
        this._updatedAt = new Date();

        this.addDomainEvent(
            new PatientAdmittedEvent(this.id, this.patientId)
        );
    }

    addNote(props: Omit<ClinicalNoteProps, 'encounterId'>): ClinicalNote {
        if (this._status === EncounterStatus.COMPLETED || this._status === EncounterStatus.CANCELLED) {
            throw new Error('Cannot add notes to a completed or cancelled encounter');
        }

        const note = ClinicalNote.create({
            ...props,
            encounterId: this.id,
        });

        this._clinicalNotes.push(note);
        this._updatedAt = new Date();
        return note;
    }

    addDiagnosis(diagnosis: Diagnosis): void {
        if (this._status === EncounterStatus.COMPLETED || this._status === EncounterStatus.CANCELLED) {
            throw new Error('Cannot add diagnoses to a completed or cancelled encounter');
        }

        const exists = this._diagnoses.some(d => d.equals(diagnosis));
        if (exists) {
            throw new Error(`Diagnosis ${diagnosis.getCode()} already exists on this encounter`);
        }

        this._diagnoses.push(diagnosis);
        this._updatedAt = new Date();
    }

    discharge(): void {
        if (this._status !== EncounterStatus.IN_PROGRESS) {
            throw new Error('Only in-progress encounters can be discharged');
        }

        this._status = EncounterStatus.COMPLETED;
        this._endDate = new Date();
        this._updatedAt = new Date();

        this.addDomainEvent(
            new PatientDischargedEvent(this.id, this.patientId, this._endDate)
        );
    }

    cancel(): void {
        if (this._status === EncounterStatus.COMPLETED) {
            throw new Error('Completed encounters cannot be cancelled');
        }

        this._status = EncounterStatus.CANCELLED;
        this._updatedAt = new Date();
    }

    // ── Domain events ──

    private addDomainEvent(event: DomainEvent): void {
        this._domainEvents.push(event);
    }

    pullDomainEvents(): DomainEvent[] {
        const events = [...this._domainEvents];
        this._domainEvents = [];
        return events;
    }
}

File: services/clinical-service/src/main/java/com/medtrust/clinical/domain/model/Encounter.java
package com.medtrust.clinical.domain.model;

import com.medtrust.clinical.domain.event.DomainEvent;
import com.medtrust.clinical.domain.event.EncounterCreatedEvent;
import com.medtrust.clinical.domain.event.PatientAdmittedEvent;
import com.medtrust.clinical.domain.event.PatientDischargedEvent;
import com.medtrust.clinical.domain.model.valueobject.Diagnosis;

import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.UUID;

public class Encounter {

    private final String id;
    private final String patientId;
    private final List<ClinicalNote> clinicalNotes;
    private final List<Diagnosis> diagnoses;
    private EncounterStatus status;
    private final Instant startDate;
    private Instant endDate;
    private final Instant createdAt;
    private Instant updatedAt;

    private final List<DomainEvent> domainEvents = new ArrayList<>();

    private Encounter(String id, String patientId, List<ClinicalNote> clinicalNotes,
            List<Diagnosis> diagnoses, EncounterStatus status,
            Instant startDate, Instant endDate, Instant createdAt, Instant updatedAt) {
        this.id = id;
        this.patientId = patientId;
        this.clinicalNotes = new ArrayList<>(clinicalNotes);
        this.diagnoses = new ArrayList<>(diagnoses);
        this.status = status;
        this.startDate = startDate;
        this.endDate = endDate;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
    }

    public static Encounter create(String patientId) {
        var encounter = new Encounter(
                UUID.randomUUID().toString(),
                patientId,
                new ArrayList<>(),
                new ArrayList<>(),
                EncounterStatus.PLANNED,
                Instant.now(),
                null,
                Instant.now(),
                Instant.now());
        encounter.addDomainEvent(new EncounterCreatedEvent(encounter.id, patientId));
        return encounter;
    }

    public static Encounter reconstitute(String id, String patientId,
            List<ClinicalNote> clinicalNotes,
            List<Diagnosis> diagnoses,
            EncounterStatus status,
            Instant startDate, Instant endDate,
            Instant createdAt, Instant updatedAt) {
        return new Encounter(id, patientId, clinicalNotes, diagnoses,
                status, startDate, endDate, createdAt, updatedAt);
    }

    // ── Domain behaviour ──

    public void admit() {
        if (this.status != EncounterStatus.PLANNED) {
            throw new IllegalStateException("Only planned encounters can be admitted");
        }
        this.status = EncounterStatus.IN_PROGRESS;
        this.updatedAt = Instant.now();
        addDomainEvent(new PatientAdmittedEvent(this.id, this.patientId));
    }

    public ClinicalNote addNote(String content, String authorId, NoteType noteType) {
        if (this.status == EncounterStatus.COMPLETED || this.status == EncounterStatus.CANCELLED) {
            throw new IllegalStateException("Cannot add notes to a completed or cancelled encounter");
        }
        var note = ClinicalNote.create(this.id, content, authorId, noteType);
        this.clinicalNotes.add(note);
        this.updatedAt = Instant.now();
        return note;
    }

    public void addDiagnosis(Diagnosis diagnosis) {
        if (this.status == EncounterStatus.COMPLETED || this.status == EncounterStatus.CANCELLED) {
            throw new IllegalStateException("Cannot add diagnoses to a completed or cancelled encounter");
        }
        boolean exists = this.diagnoses.stream().anyMatch(d -> d.equals(diagnosis));
        if (exists) {
            throw new IllegalArgumentException(
                    "Diagnosis " + diagnosis.getCode() + " already exists on this encounter");
        }
        this.diagnoses.add(diagnosis);
        this.updatedAt = Instant.now();
    }

    public void discharge() {
        if (this.status != EncounterStatus.IN_PROGRESS) {
            throw new IllegalStateException("Only in-progress encounters can be discharged");
        }
        this.status = EncounterStatus.COMPLETED;
        this.endDate = Instant.now();
        this.updatedAt = Instant.now();
        addDomainEvent(new PatientDischargedEvent(this.id, this.patientId, this.endDate));
    }

    public void cancel() {
        if (this.status == EncounterStatus.COMPLETED) {
            throw new IllegalStateException("Completed encounters cannot be cancelled");
        }
        this.status = EncounterStatus.CANCELLED;
        this.updatedAt = Instant.now();
    }

    // ── Domain events ──

    private void addDomainEvent(DomainEvent event) {
        this.domainEvents.add(event);
    }

    public List<DomainEvent> pullDomainEvents() {
        var events = List.copyOf(this.domainEvents);
        this.domainEvents.clear();
        return events;
    }

    // ── Getters ──

    public String getId() {
        return id;
    }

    public String getPatientId() {
        return patientId;
    }

    public List<ClinicalNote> getClinicalNotes() {
        return Collections.unmodifiableList(clinicalNotes);
    }

    public List<Diagnosis> getDiagnoses() {
        return Collections.unmodifiableList(diagnoses);
    }

    public EncounterStatus getStatus() {
        return status;
    }

    public Instant getStartDate() {
        return startDate;
    }

    public Instant getEndDate() {
        return endDate;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getUpdatedAt() {
        return updatedAt;
    }
}
